log connection info to Elasticsearch

msimerson opened the pull request:
### change history
* merge transaction results into connection results
* get decoded message headers
* skip logging connection if already logged in transaction
* added exclude_hosts option
    handy for not polluting ES results with monitoring and testing connections
* add default config file
* move transaction "stored" node outside cb
    so that connections doesn't get stored if the transaction already did
* make ES index names configurable
* added config setting to nest plugin result inside another object
* add tests
* added log_connections option
* list of message headers to add is now a config setting
* the top level names (plugin, connection, message) are configuration with config options
* connection properties names can be aliased with config settings in [report_as] block
* removed index map template (until the docs for the ES API I'm using exist, and I can figure out how to detect failures and report 'em, it's better to have a static map)
* new top level `message` object.
* list of connection properties to store is now configurable
* added auth info

msimerson commented:
Needs docs and tests, but apparently works decent for at least two people.

> 3:17 iksik: hm, some time ago i've noticed You guys were talking about pushing haraka logs into elasticsearch... is there any plugin i can use for that? or the only way is to hack something custom?
> 3:17 ultimatt: https://github.com/baudehlo/Haraka/pull/906
> 3:18 iksik: awesome, thank You ;)
> 3:19 ultimatt: report back in that PR
> 4:25 iksik: ultimatt: it seems it works well - for now ;-)

msimerson commented:
I've updated this plugin and [sample Elasticsearch output is here](https://gist.github.com/msimerson/7d0011b085a0ac06ccdf)

celesteking commented:
Except that it doesn't log what's gonna happen after disconnect, like smtp_forward/proxy mail delivery (if ever successful). What are you thoughts about this?

smfreegard commented:
smtp_forward and smtp_proxy all happen before disconnect and before the end of the transaction.   It's only outbound that queues and runs afterwards.   At that point you'd have to use the delivered, deferred and bounce hooks.

msimerson commented:
> Except that it doesn't log what's gonna happen after disconnect, like smtp_forward/proxy mail delivery (if ever successful). What are you thoughts about this?

As @smfreegard said, it's already addressed. This is from a logged ES transaction on my server:

```json
"queue/smtp_forward": {
      "pass": [
        [
          "ok 1432749543 qp 31485"
        ]
      ]
    }
```

However, that being there depends on #962, which has smtp_forward store it's results. When the queue operation is successful, the results here are the same as listening on queue_ok hook. When anything else happens during smtp_forward (rejected, connection drop, timeout), this is the place where it shows up.

celesteking commented:
Where's the config file sample for this plugin?

msimerson commented:
Just added.

celesteking commented:
Thanks. Why the reason for 2 index names? Kibana can't combine them in 1 search.

msimerson commented:
Temporary. I should go ahead and merge them now.

msimerson commented:
> Why the reason for 2 index names?

I remember now. Connections and transactions are indexed separately because they are very different things. Even within the same connection, transactions are not necessarily related to each other. Each connection can have zero or many transactions, with each transaction representing a single message with one-to-many recipients. Nesting transactions within an ES connection was going to be a pretty horrible data model. Instead, each transaction is saved as its own ES document. In practice, this works out splendidly, and in most cases, the transactions are the only "interesting" data.

The **only** time a connection is logged is when it has zero transactions. If you search in the log file for the string 'skipping already logged txn', you'll see where I skip saving connections if they've already been saved to ES. In practice, what this means is that the only thing in the connections index is clients that don't initiate transactions (monitoring apps, etc..) and connections that have been denied for policy reasons, before they initiate a transaction. As such, they're 'connections', and not 'transactions'. In practice, at least in my case, they're all noise and no signal.

I believe this is the correct default as IIRC, @smfreegard doesn't bother saving the connections at all. If you have a compelling reason for mixing them in with the transactions, I'd be interested to hear it, and willing to accept a PR that adds a config file option for dis/enabling that.

celesteking commented:
okay, looks good and point taken. 
What worries me is plugin results not separated from other cruft. Would be wonderful to split that off for readability. 

Most of our support requests concern mail not being sent/received, so having data on why something was denied is top priority. Lowest priority is getting overall usage stats. The so-so priority is having notifications about spam attempts.  These are our use cases.

msimerson commented:
> What worries me is plugin results not separated from other cruft. Would be wonderful to split that off for readability.

I think having them split off actually harms readability and kibana usability. However, it's really easy to change if you want to try it that way:

```diff
--- plugins/log.elasticsearch.js	2015-05-28 19:37:40.044889942 -0700
+++ foo	2015-06-02 10:55:53.935116613 -0700
@@ -64,7 +64,9 @@
         if (plugin.cfg.ignore_hosts[connection.remote_host]) return next();
     }
 
-    var res = plugin.get_plugin_results(connection);
+    var res = {
+        plugins: plugin.get_plugin_results(connection),
+    };
     res.timestamp = new Date().toISOString();
     res.txn = {
         mail_from: connection.transaction.mail_from.original,
```

I could make nesting the plugin results a config option.

msimerson commented:
Updated, with these changes:

* index names are now configurable
* added config setting to nest plugin result inside another object
* added log_connections option
* added documentation for config settings
* added tests
* automatically apply the index map template (unless disabled)

msimerson commented:
### adding more properties && rearranging the JSON

* list of message headers to add is now a config setting
* the top level names (plugin, connection, message) are configuration with config options
* connection properties names can be aliased with config settings in [report_as] block
* new top level `message` object. Structure now looks like this:
    * connection
        * local
            * ip
            * port
            * host
        * remote
            * ip
            * port
            * host
        * hello
            * verb
            * host
        * count
        * relay
        * other connection properties, per .ini
    * message
        * bytes
        * envelope
            * sender
            * recipient
        * header
            * To / From / Subject (per config settings)
        * body
            * attachment
                * file
                * ctype
                * md5
                * bytes
        * queue
    * plugin
        * asn
        * karma
        * geoip
        * spamassassin
        * ....

* improved structure makes index map simpler 
    * all fields named 'ip' are type 'ip'
    * all fields named 'host', etc....

msimerson commented:
### Example ES Document

```json
{
  "timestamp": "2015-06-13T22:47:36.472Z",
  "conn": {
    "local": {
      "ip": "127.0.0.30",
      "port": 25,
      "host": "node.tnpi.net"
    },
    "remote": {
      "ip": "158.85.175.54",
      "host": "158.85.175.54-static.reverse.softlayer.com",
      "port": 53660
    },
    "hello": {
      "host": "ns.lomce.com",
      "verb": "EHLO"
    },
    "count": {
      "errors": 0,
      "msg": {
        "accept": 0,
        "tempfail": 0,
        "reject": 1
      },
      "rcpt": {
        "accept": 1,
        "tempfail": 0,
        "reject": 0
      },
      "trans": 1
    },
    "tls": false,
    "relay": false,
    "bytes": 329779,
    "pipelining": true,
    "early": false,
    "duration": 48.847
  },
  "message": {
    "bytes": 329779,
    "envelope": {
      "sender": "inland_commission@outlook.com",
      "recipient": [
        {
          "action": "accept",
          "address": "j****e@si******n.net"
        }
      ]
    },
    "header": {
      "From": "\"Inland Commission\"<inland_commission@outlook.com>",
      "Subject": "Payment Received!"
    },
    "body": {
      "attachment": [
        {
          "file": "Payment Confirmation.zip",
          "ctype": "application/x-zip-compressed",
          "md5": "f40149b6323d00ba87889f32b680467f",
          "bytes": 238194
        }
      ]
    },
    "queue": {}
  },
  "plugin": {
    "karma": {
      "fail": [
        "fcrdns.fail",
        "dnsbl.fail",
        "dnsbl.fail",
        "dnsbl.fail",
        "dnsbl.fail",
        "dnsbl.fail",
        "asn.fail",
        "spf.result",
        "deny:attachment"
      ],
      "score": -29,
      "good": "0",
      "bad": "2",
      "connections": "3",
      "history": -2
    },
    "relay": {
      "msg": [
        "acl(unlisted)"
      ]
    },
    "access": {
      "msg": [
        "unlisted(mail)",
        "unlisted(rcpt)"
      ]
    },
    "dnsbl": {
      "fail": [
        "psbl.surriel.com",
        "dnsbl-1.uceprotect.net",
        "b.barracudacentral.org",
        "bl.spamcop.net",
        "dnsbl.justspam.org",
        "dnsbl.sorbs.net"
      ]
    },
    "limit": {
      "pass": [
        "1/1"
      ]
    },
    "spf": {
      "scope": "mfrom",
      "result": "SoftFail",
      "domain": "outlook.com"
    },
    "avg": {
      "pass": [
        "clean"
      ]
    },
    "p0f": {
      "first_seen": 1434235607,
      "last_seen": 1434235607,
      "total_conn": 1,
      "distance": 15,
      "os_name": "Linux",
      "os_flavor": "3.11 and newer",
      "link_type": "DSL"
    },
    "fcrdns": {
      "pass": [
        "fcrdns"
      ],
      "fail": [
        "is_generic_rdns"
      ],
      "fcrdns": [
        "158.85.175.54-static.reverse.softlayer.com"
      ],
      "ptr_names": [
        "158.85.175.54-static.reverse.softlayer.com"
      ],
      "ptr_multidomain": false,
      "has_rdns": true,
      "ptr_name_has_ips": true,
      "ptr_name_to_ip": [
        {
          "k": "158.85.175.54-static.reverse.softlayer.com",
          "v": [
            "158.85.175.54"
          ]
        }
      ]
    },
    "asn": {
      "fail": [
        "karma"
      ],
      "asn": "36351",
      "org": "SoftLayer Technologies Inc.",
      "asn_score": -364,
      "asn_connections": "506",
      "asn_good": "31",
      "asn_bad": "395"
    },
    "geoip": {
      "continent": "NA",
      "country": "US",
      "region": "TX",
      "city": "Dallas",
      "ll": [
        32.9299,
        -96.8353
      ],
      "distance": "19"
    },
    "uribl": {},
    "helo": {
      "pass": [
        "valid_hostname",
        "match_re",
        "bare_ip",
        "dynamic",
        "big_co(not)",
        "literal_mismatch"
      ],
      "multi": true
    },
    "bounce": {
      "isa": "no"
    },
    "attachment": {},
    "clamd": {
      "fail": [
        "Sanesecurity.Malware.22041.ZipHeur.UNOFFICIAL"
      ]
    },
    "rspamd": {
      "is_spam": true,
      "is_skipped": false,
      "score": 20.5,
      "required_score": 15,
      "action": "reject",
      "RBL_UCEPROTECT_LEVEL1": 1,
      "MISSING_MID": 3,
      "RBL_SORBS_RECENT": 1.5,
      "R_UNDISC_RCPT": 5,
      "MISSING_TO": 2,
      "R_NO_SPACE_IN_FROM": 1,
      "FORGED_MUA_OUTLOOK": 3,
      "FORGED_RECIPIENTS": 1,
      "RECEIVED_SPAMHAUS_XBL": 3,
      "messages": "(SPF): spf softfail",
      "time": 0.231
    },
    "spamassassin": {
      "headers": {
        "ASN": "AS36351 158.85.160.0/19",
        "Report": "\r\n\t*  2.7 RCVD_IN_PSBL RBL: Received via a relay in PSBL\r\n\t*      [158.85.175.54 listed in psbl.surriel.com]\r\n\t*  0.0 NSL_RCVD_FROM_USER Received from User\r\n\t*  0.7 SPF_SOFTFAIL SPF: sender does not match SPF record (softfail)\r\n\t*  1.6 DATE_IN_PAST_03_06 Date: is 3 to 6 hours before Received: date\r\n\t*  0.0 FREEMAIL_FROM Sender email is commonly abused enduser mail provider\r\n\t*      (inland_commission[at]outlook.com)\r\n\t*  1.0 MISSING_HEADERS Missing To: header\r\n\t*  2.0 BAYES_80 BODY: Bayes spam probability is 80 to 95%\r\n\t*      [score: 0.9335]\r\n\t*  1.4 RCVD_IN_BRBL_LASTEXT RBL: No description available.\r\n\t*      [158.85.175.54 listed in bb.barracudacentral.org]\r\n\t*  1.1 DCC_CHECK Detected as bulk mail by DCC (dcc-servers.net)\r\n\t*  1.3 RCVD_IN_BL_SPAMCOP_NET RBL: Received via a relay in bl.spamcop.net\r\n\t*      [Blocked - see <http://www.spamcop.net/bl.shtml?158.85.175.54>]\r\n\t*  1.5 FSL_RCVD_IN_XBL_DEEP Received via a relay in Spamhaus XBL (not\r\n\t*      lastexternal)\r\n\t*  0.5 MISSING_MID Missing Message-Id: header\r\n\t*  1.6 REPLYTO_WITHOUT_TO_CC No description available.\r\n\t*  0.0 FROM_MISSP_MSFT From misspaced + supposed Microsoft tool\r\n\t*  2.7 MALFORMED_FREEMAIL Bad headers on message from free email service\r\n\t*  0.0 FROM_MISSP_USER From misspaced, from \"User\"\r\n\t*  0.0 AXB_XMAILER_MIMEOLE_OL_024C2 Yet another X header trait\r\n\t*  1.4 FSL_NEW_HELO_USER Spam's using Helo and User\r\n\t*  0.0 FROM_MISSP_XPRIO Misspaced FROM + X-Priority\r\n\t*  2.5 TO_NO_BRKTS_FROM_MSSP Multiple formatting errors\r\n\t*  0.0 FROM_MISSP_REPLYTO From misspaced, has Reply-To\r\n\t*  0.0 FROM_MISSPACED From: missing whitespace\r\n\t*  1.9 FORGED_MUA_OUTLOOK Forged mail pretending to be from MS Outlook\r\n\t*  1.1 FROM_MISSP_EH_MATCH From misspaced, matches envelope\r\n\t*  2.5 TO_NO_BRKTS_MSFT To: misformatted and supposed Microsoft tool\r\n\t*  0.0 FROM_MISSP_FREEMAIL From misspaced + freemail provider",
        "Status": "Yes, score=27.5 required=5.0 autolearn=no autolearn_force=no",
        "DCC": "dmv.com: spamassassin.tnpi.net 1181; Body=1 Fuz1=7 Fuz2=many",
        "Checker-Version": "SpamAssassin 3.4.1 (2015-04-28) on\r\n\tspamassassin.tnpi.net"
      },
      "flag": "Yes",
      "score": "27.5",
      "reqd": "5.0",
      "tests": "AXB_XMAILER_MIMEOLE_OL_024C2,BAYES_80,DATE_IN_PAST_03_06,\r\n\tDCC_CHECK,FORGED_MUA_OUTLOOK,FREEMAIL_FROM,FROM_MISSPACED,FROM_MISSP_EH_MATCH,\r\n\tFROM_MISSP_FREEMAIL,FROM_MISSP_MSFT,FROM_MISSP_REPLYTO,FROM_MISSP_USER,\r\n\tFROM_MISSP_XPRIO,FSL_NEW_HELO_USER,FSL_RCVD_IN_XBL_DEEP,MALFORMED_FREEMAIL,\r\n\tMISSING_HEADERS,MISSING_MID,NSL_RCVD_FROM_USER,RCVD_IN_BL_SPAMCOP_NET,\r\n\tRCVD_IN_BRBL_LASTEXT,RCVD_IN_PSBL,REPLYTO_WITHOUT_TO_CC,SPF_SOFTFAIL,\r\n\tTO_NO_BRKTS_FROM_MSSP,TO_NO_BRKTS_MSFT",
      "time": 0.839
    },
    "qmail_deliverable": {
      "pass": [
        "rcpt.vpopmail dir"
      ],
      "msg": [
        "mail_from.not local",
        "sock: 127.0.0.6:8998"
      ]
    },
    "is_resolvable": {
      "pass": [
        "has_a_records"
      ]
    },
    "in_host_list": {
      "msg": [
        "mail_from!local",
        "rcpt!local"
      ]
    },
    "headers": {
      "pass": [
        "from_match",
        "duplicate",
        "missing",
        "invalid_date",
        "Return-Path",
        "UA(Microsoft Ou)",
        "direct-to-mx(2)"
      ],
      "msg": [
        "not MLM"
      ]
    }
  }
}
```

msimerson commented:
Added auth results.

```js
auth = {
    pass: "auth/auth_vpopmaild",
    method: "PLAIN",
    user: "user@example.com",
    fail: [],
};
```

smfreegard commented:
This all looks good to me.  I'm reluctant to merge it yet as there have been so many changes since the pull was created!   Giving it a few more days to settle.

msimerson commented:
> Giving it a few more days to settle.

Whatever could you mean? ;-) I've only made two commits since your comment. 

(yes, I share those sentiments. You had already reviewed it last week, and I squashed it, and **thought** about merging, but figured it better to wait until I haven't changed it for at least a week.)

msimerson commented:
A week passed without a single change! I don't expect many future changes beyond optimizations for the index map template (in the doc file).
